crypto/ec/ec_asn1.c

```diff
@@ -792,18 +794,79 @@ EC_GROUP *EC_GROUP_new_from_ecparameters(const ECPARAMETERS *params)
         goto err;
     }
 
+    /*
+     * Check if the explicit parameters group just created matches one of the
+     * built-in curves.
+     *
+     * We create a copy of the group just built, so that we can remove optional
+     * fields for the lookup: we do this to avoid the possibility that one of
+     * the optional parameters is used to force the library into using a less
+     * performant and less secure EC_METHOD instead of the specialized one.
+     * In any case, `seed` is not really used in any computation, while a
+     * cofactor different from the one in the built-in table is just
+     * mathematically wrong anyway and should not be used.
+     */
+    if ((ctx = BN_CTX_new()) == NULL) {
+        ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, ERR_R_BN_LIB);
+        goto err;
+    }
+    if ((dup = EC_GROUP_dup(ret)) == NULL
+            || EC_GROUP_set_seed(dup, NULL, 0) != 1
+            || !EC_GROUP_set_generator(dup, point, a, NULL)) {
```

Comment:
Side note: With @bbbrumley's recent changes this presumably ends up "guessing" the cofactor anyway.

Reply:
Uhm, yes I considered it, I had 2 choices:

- `BN_zero(dup_>cofactor)`
- Let @bbbrumley magic happen

I opted for the second one because I figured that if the magic happens but the guess does not match the built-in table, it means we have an error in the code (that should be caught by @bbbrumley testing and also by the unit tests from this PR). In the end what we really want to do is remove agency from an attacker poisoning the explicit parameter encoding with a wrong cofactor (either zero or some bogus number), so guessing correctly instead of zeroing works as cofactor is not an abusable user input anyway.

Reply:
Makes sense.